Programmable Control Flow

sylvanc opened the pull request:
This change produces a much simpler AST that uses `select` and `lambda` as its fundamental pieces. There is no distinction once the AST is built between prefix, infix, or dot notation, and the built in control flow is `try`, `match`, and `when`. Control flow is built by using non-local returns, which are encoded with `try/throw`.

nwf commented on docs/type-lists.md:
````diff
@@ -0,0 +1,73 @@
+```ts
+papply[A, B..., C, R: B->C](f: R~>(A, B...)->C, x: R~>A): R
+{
+  { y: B... => f(x, y) }
````
May be important to indicate that tuple flattening also happens at the value level as well (i.e., this is not `(x, (y))`)?  And that this flattening is type-directed, based on the fact that `y : B...`.  If instead `z : (B...)` or `z : T` and `T ~ (B...)`, then `(x, z)` and `(x, (z))` are identical?

sylvanc replied:
Thanks, I've lifted you language!

nwf commented on docs/unified-call-syntax.md:
```diff
@@ -0,0 +1,192 @@
+
+An unqualified function name `identifier` (as distinct from a qualified name such as `ident1::ident2`) can be used. If there is a function in the current scope (either defined in the current class or made available without qualification with a `using` directive), this will resolve as if it were a fully-qualified reference.
+
+If there is no static resolution to a function, this will instead resolve as if it were a dynamic unbound selection on the first argument. It's a compile-time error to try to use an undefined function with no arguments.
```
"unqualified", not "undefined"?

nwf commented on docs/unified-call-syntax.md:
```diff
@@ -0,0 +1,192 @@
+let f = foo;
+
+// If `foo` is in scope, this calls `foo(x)`. If `foo` is not in scope, this
+// calls `x::foo(x)`.
```
"this behaves as `x.foo()`, that is, it calls `x::foo(x)`"?

nwf commented on docs/unified-call-syntax.md:
```diff
@@ -0,0 +1,192 @@
+a f b c;
+
+// If the second argument is a tuple, nest it.
+f(a, (b, c));
```
This isn't `a f (b,c)` by the above rules (because that's `f(a,b,c)`)... but is it `a f ((b,c))`?

The examples so far don't, I think, distinguish between syntactic and type-based operation.  That is, is `let bc = (b,c) in a f bc` the same as `f(a,b,c)` or `f(a,(b,c))`?

nwf commented on docs/unified-call-syntax.md:
```diff
@@ -0,0 +1,192 @@
+
+## Variable Length Arguments
+
+There is no language support for variable length arguments. Instead, use an iterator.
```
Perhaps something like "(Recall that type lists, despite their apparent variable length, elaborate into concrete tuples of particular lengths)"?

sylvanc commented on docs/type-lists.md:
```diff
@@ -0,0 +1,73 @@
+# Type Lists
+
+`T...` means a type list called `T...`. In a generic, if it is given a constraint, the constraint applies to each type in the list, not the list itself. So `[T...: U]` means a type list of any length such that each element is a subtype of `U`. The type `()` is an empty type list. Unlike a singleton type, this is not a subtype of the empty interface.
```
Yes, thanks!

sylvanc commented on docs/type-lists.md:
````diff
@@ -0,0 +1,73 @@
+Parameters are tuples that may contain type lists. Calling a function does a destructuring bind of the tuple argument. Results are tuples that may contain type lists. Using the result also does a destructuring bind.
+
+```ts
+papply[A, B..., C, R: B->C](f: R~>(A, B...)->C, x: R~>A): R
````
Nice catch! It's the latter, so this needs fixing.

sylvanc commented on docs/type-lists.md:
```diff
@@ -0,0 +1,73 @@
+
+`T...` means a type list called `T...`. In a generic, if it is given a constraint, the constraint applies to each type in the list, not the list itself. So `[T...: U]` means a type list of any length such that each element is a subtype of `U`. The type `()` is an empty type list. Unlike a singleton type, this is not a subtype of the empty interface.
+
+`A...` is a tuple of zero or more types. `(A, B...)` is a tuple of one or more types. That is, the type list is flattened into the tuple. `A...->B` is a function type with zero or more arguments. `(A, B...)->C` is a function type with one or more arguments.
```
Yes. And if `B...` is bound to a tuple of length 1, then `(A, B...)` and `(A, (B...))` are identical.

sylvanc commented on docs/type-lists.md:
```diff
@@ -0,0 +1,73 @@
+
+`A...-->B` is a curried function type (note the use of `-->` instead of `->`). That is, the type list is flattened into the arrow. If `A...` is bound to `(C, D)`, you would get `C->D->B`.
+
+Parameters are tuples that may contain type lists. Calling a function does a destructuring bind of the tuple argument. Results are tuples that may contain type lists. Using the result also does a destructuring bind.
```
Thanks, I lifted your language here too!

sylvanc commented:
Thanks @nwf ! I've incorporated your changes in the type lists doc, and rewritten the UCS doc completely. If you have a chance, would you mind reading the UCS doc again?

nwf commented on docs/programmable-control-flow.md:
```diff
@@ -0,0 +1,145 @@
+A match that doesn't match anything throws `NoMatch`.
+`match` can't be written in the source language because we want to detect exhaustive matches.
+
+> TODO: destructure in patterns
```
FWIW, the sort of analogy in Haskell is View Patterns, which don't project a `Bool` like `==` here, but project an arbitrary value which is then subjected to the usual pattern handling (I think including additional views, though I'm not sure why you'd do that).  So you'll see a lot of `view :: ... -> Maybe ...` views that get used like `foo (view x -> Just z) = ... z ...`.  (see https://gitlab.haskell.org/ghc/ghc/-/wikis/view-patterns)

sylvanc replied:
I'm definitely not following this. I'll bug you for a walkthrough.

nwf commented on docs/programmable-control-flow.md:
```diff
@@ -0,0 +1,145 @@
+
+New language features:
+
+* Lambda parameters can be values, which turns the lambda into a case.
```
What's the rule for deciding whether a parameter is a value or a variable binder?  That is, when is `{ x => ... }` testing equality against `x` and when is it introducing `x`?

sylvanc replied:
I've added a note here. The rule is: If it can be a parameter, it is. Otherwise, it is interpreted as a pattern.

nwf commented on docs/type-lists.md:
```diff
@@ -0,0 +1,75 @@
+
+Function parameters and results may have type annotations containing type lists. Calling a function does a destructuring bind of the tuple argument. Using the result also does a destructuring bind.
+
+Tuple flattening also happens at the value level. This flattening is type-directed: given `x: A...`, uses of `x` are expanded to a sequence of values within the enclosing tuple. For example, given `x: A, y: B...`, if `B...` is bound to a tuple of length 2, `(x, y)` is a tuple of length 3. If instead `B...` is bound to a tuple of length 1, then `(x, y)` and `(x, (y))` are identical.
```
Perhaps also add to the "given `x: A, y: B...`, if `B...` is bound to a tuple of length 2," case that `(x, (y))` is not `(x, y)` but rather a pair whose second component is a pair?

sylvanc replied:
Good call.

nwf commented on docs/type-lists.md:
```diff
@@ -0,0 +1,75 @@
+  { x: A =>
+    match f
+    {
+      { g: R~>A->C => papply(g, x) }
```
```suggestion
      { g: R~>(A->C) => papply(g, x) }
```

nwf commented on docs/type-lists.md:
```diff
@@ -0,0 +1,75 @@
+  { y: B... => f(x, y) }
+}
+
+curry_one[A, B..., C, R: A->B...->C](f: R~>(A, B...)->C): R
```
```suggestion
curry_one[A, B..., C, R: A->B...->C](f: R~>((A, B...)->C)): R
```

nwf commented on docs/type-lists.md:
```diff
@@ -0,0 +1,75 @@
+  { x: A, y: B... => f x y }
+}
+
+curry_all[A, B..., C, R: A->B...-->C](f: R~>(A, B...)->C): R
```
```suggestion
curry_all[A, B..., C, R: A->B...-->C](f: R~>((A, B...)->C)): R
```

nwf commented on docs/type-lists.md:
```diff
@@ -0,0 +1,75 @@
+  match x
+  {
+    { () => 0 }
+    { _, tl => 1 + length(tl) }
```
It probably deserves mention somewhere that despite the existence of type lists and their implicit tuple flattening behavior, that this flattening happens only on the right, so that there's no ambiguity about how many things get cleaved off by that `_`.  (Similarly, it might be good to say what happens---even if that's a failure to compile---if `x : A..., y : B...` in `(x, y)`.)
